olm/operator-registry.md

```diff
@@ -0,0 +1,322 @@
+---
+title: operator-registry
+authors:
+  - "@kevinrizza"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2019-09-05
+last-updated: 2019-09-20
+status: provisional
+---
+
+# operator-registry
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+The purpose of this enhancement is to create a first class storage mechanism for Operators in Kubernetes.
+
+This implementation is designed with commonly used tools in mind (specifically, container images and registries) to provide a format of pushing and pulling yaml metadata needed to install and lifecycle Operators.
+
+At a glance, the intention of this enhancement is to generate a format to define index containers that are equivalent to a "repository" and can be installed on a cluster. These indexes contain a set of references to already published Operators as well as a set of metadata that allows users to discover what can be installed on their cluster and how to install it (selecting a version, understanding what dependencies are required, etc.).
+
+Additionally, once those index containers are defined, this enhancement also intends to drive the OperatorHub/OLM user experience with the metadata in those indexes.
+
+## Motivation
+
+The primary motivation for this enhancement is to replace the current implementation of streaming catalog data from app-registry bundles that exist in OpenShift. These bundles exist as base64 encoded file directories that are stored and referenced by an API. Currently, this is only implemented as a closed source API on Quay.io.
+
+Additionally, creating a scaleable solution that fits the needs of operator developers, cluster owners, and development pipelines is important.
+
+### Goals
+
+  - Have index containers that drive on cluster catalog metadata.
+  - Index containers define what Operators can be installed on cluster and how to get the manifests to install them.
+  - Drive the Console (embedded Operator Hub) from the indexes
+  - On clicking install / creating subscription pull down necessary operator manifest containers + all dependent manifest containers to drive metadata for installation
+  - A way to serve that API so that existing apis can be driven by the metadata once it is unpacked
+  - Index catalogs can stream updates from external source
+  - Remove the need to constantly pull down all metadata associated with the entire catalog on each update to any element of the catalog -- this will also allow the update poll timer to be configurable
+
+### Non-Goals
+
+  - Tools to build the index images or operator manifest images
```

Comment:
Providing this tool somehow will be essential to success here.  It's doesn't need to be a goal of this enhancement, but it needs to be taken up if this is all to work.

Reply:
We have a separate enhancement specifically for tooling (though some will be built as part of this enhancement).